### ISSUE TITLE
Hashes/encodings below the heuristic limit are treated as typos

### ISSUE THREAD

**Issue by halkeye** [crate-ci/typos#415]
```
error: `Ba` should be `By`, `Be`
  --> ./content/blog/2017/08/2017-08-08-introducing-jenkins-minute.adoc:39:14
   |
39 | video::FhDomw6BaHU[youtube, width=852, height=480]
```

I can add it to `[default.extend-identifiers]` so its not a blocker but figured you'd like another test case.

**Comment by epage** [comment]
The challenge is being able to identify that has a hash.  How do we tell a hash from an identifier?

Right now, we support
- SHA detection: must be 32+ characters long and consistent case
- Base 64 detection: Must be 90 characters long or have `+` / `/` in it and must have the padding bytes (though there is uncertainty if the padding byte requirement will stay, see #413)

**Comment by halkeye** [comment]
yea that makes a lot of sense. I'm just starting to use the app and loving it so far, I've only had to whitelist two hashes that have ba in them so its not a big deal for me.

Maybe some sort of regex or something so I could whitelist `video::[a-zA-Z0-9]\[`

**Comment by tspearconquest** [comment]
Hello,

Git commit hashes tend to run in the range `[0-9a-fA-F]{7,}` so that would be a useful addition.

**Comment by epage** [comment]
@tspearconquest for shorter git commit hashes, we'll need to rely on a heuristic like talked about in #484 because shorter commit hashes could just as easily be words.

**Comment by jplatte** [comment]
How about adding a heuristic "word contains characters preceded by numbers" (where "word" is a whitespace-separated segment, not a case-separated segment)? I don't think I've ever seen an identitifer be named `foo1bar` or `3foo`, though `foo3` or `foo3_bar` seem realistic (e.g. `zip3`, `zip4`).

**Comment by halkeye** [comment]
sha1hash?

**Comment by jplatte** [comment]
Right, there's a few exceptions (I also remembered there being `2to3`), but maybe it's still a good heuristic? Personally, I consider false positives a bigger issue than false negatives, and I think that matches typos' overall approach.

**Comment by pums974** [comment]
There is also all the this2that and thing4stuff

**Comment by epage** [comment]
@jplatte I'd probably refine your comment to be "any identifier that exclusively word splits due to numbers and not any other separator (be it case or `_`)

The next question is the likelihood of a shortened sha having no numbers.  I probably didn't bring this up in the other thread talking about heuristics but I suspect to have something always complain than it have it complain in a way people no longer expect.

**Comment by jplatte** [comment]
In the case of a hex string, that would be `(10/16)` (since 6 out of the 16 possible chars are alphabetical) to the power of the string length. git short hashes are the shortest hash I see in practice, and they seem to start at 7 characters (longer in large repos), which puts the probability of such a hash having no digits at pretty much exactly 0.1%. Strings that contain no digits *before any non-digit characters* would be closer to 0.5% though (rough estimation, could also be >0.5%, but not <0.28%).

**Comment by epage** [comment]
FYI #695 provides a new workaround for false positives

**Comment by azzamsa** [comment]
```
  --> ./content/n/rust-docker.md:52:44
   |
52 | hello         0.1.0                ac4e1a72ba05   2 minutes ago    1.38GB
   |                                            ^^
   |
error: `ba` should be `by`, `be`
  --> ./content/n/rust-docker.md:53:46
   |
53 | rust          1.52.1-slim-buster   61cb3c65a6ba   3 weeks ago      621MB
   |                                              ^^
 ```
 
 The `extend-ignore-re` solved this issue.
 
 ```
 [default]
extend-ignore-re = ["[0-9a-fA-F]{12}"]
```

I think we can safely close this issue.

**Comment by jplatte** [comment]
@azzamsa that regex is much too generic, it disables spell-checking for all 12-letter identifiers as well.

**Comment by DeadNews** [comment]
> ## Description
> 
> The `typos` pre-commit hook fails on truncated commit hashes in `CHANGELOG.md`.
> ## Environment
> 
> ```yaml
> - repo: https://github.com/crate-ci/typos
>   rev: v1.20.4
>   hooks:
>     - id: typos
> ```
> 
> ## Actual Behavior
> 
> ```
> $ pre-commit run --files CHANGELOG.md
> 
> typos....................................................................Failed
> - hook id: typos
> - exit code: 2
> 
> error: `ba` should be `by`, `be`
>   --> CHANGELOG.md:100:28
>     |
> 100 | - _(README)_ update - ([e84ba3e](https://github.com/DeadNews/firebirdsql-run/commit/e84ba3e8e2f72a8dcad43f8ac3c768527ca199bd))
>     |                            ^^
>     |
> ```
> 
> ```
> $ pre-commit run --files CHANGELOG.md
> 
> typos....................................................................Failed
> - hook id: typos
> - exit code: 2
> 
> error: `ba` should be `by`, `be`
>   --> CHANGELOG.md:22:99
>    |
> 22 | - update `mkdocs` config ([#127](https://github.com/DeadNews/encode-utils-cli/issues/127)) - ([c92ba20](https://github.com/DeadNews/encode-utils-cli/commit/c92ba2032ac0b492b390d45c50f7c57c2660df5c))
>    |                                                                                                   ^^
>    |
> error: `ba` should be `by`, `be`
>   --> CHANGELOG.md:62:43
>    |
> 62 | - _(renovate)_ use shared config - ([693c3ba](https://github.com/DeadNews/encode-utils-cli/commit/693c3ba58822db45dd06a032ba1ce554db6deaf6))
>    |                                           ^^
>    |
> ```

original: https://github.com/crate-ci/typos/issues/982

**Comment by DeadNews** [comment]
>`ba` should be `by`, `be`

↑ This `ba` is in all examples. 
Maybe add it to the exceptions?

**Comment by dmadisetti** [comment]
Example of SHA-256 hash "typo"

```text
typos tests/_save/test_hash.py                                                                                                                                                                                                                
error: `iy` should be `it`
  --> tests/_save/test_hash.py:785:51
    |
785 |             expected_hash = "rC6YiNsuaZKQ1JqMgSRa0iyEDwi7ZTl6InABjuM0RDY"
    |                                                   ^^
    |
```

Using `typos-cli 1.24.5`.

Seems like defaults might be a little loose

**Comment by mathause** [comment]
Concerning the `"id": "ba638183"` in jupyter notebooks I had success in ignoring them with:

```toml
[type.jupyter] # or [tool.typos.type.jupyter] in pyproject.toml
extend-ignore-re = [
    "\"id\": \".*\"",
]
```

**Comment by CathalMullan** [comment]
Would be nice to catch [SRI hashes](https://www.w3.org/TR/sri) specifically.
Shouldn't result in many false positives.

Example of a miss from the nixpkgs repo:

```
error: `Nwe` should be `New`
  --> ./pkgs/by-name/co/colorpanes/package.nix:19:32
   |
19 |   cargoHash = "sha256-+ltcTuLksNwe7KIt8apYNZkMoA2w4EObG5dhJliRb6Y=";
   |                                ^^^
```

```
sha256-   abc123def456ghi789jkl012mno345pqr678stu901vwxy   =
^         ^                                                ^
static    43 base64 digits                                 1 base64 padding digit

sha384-   xyz789uvw012abc345def678ghi901jkl234mno567pqr890stu123vwx456yzab
^         ^
static    64 base64 digits

sha512-   pqr345stu678vwx901yza234bcd567efg890hij123klm456nop789qrs012tuv345wxy678z   ==
^         ^                                                                           ^
static    86 base64 digits                                                            2 base64 padding digits
```
